### DISCUSSION TITLE
A subroutine example

### DISCUSSION THREAD

**Discussion by eltikia** [eltikia/Perl-1.0#41] [Q&A] [answered]
Can you give us an example of a subroutine in Perl? I'm studying for a Java exam and I'd like to know more. Here's a quotation from a handbook.
> Different languages handle parameters in different ways. Pass-by-value is used by many languages, including Java. [...] The other approach is pass-by-reference. It is used by default in a few languages, such as Perl.

**Comment by Stagyrite** [comment, marked as answer]
Let's consider what the handbook says, and we need both Perl 1 and Perl 5 interpreters to find out how it works.

### No call-by-reference in Perl 1
A pass-by-reference example wouldn't work in Perl 1, and here is a good try.
```
sub trySwap {
    $temp = $_[0];
    $_[0] = $_[1];
    $_[1] = $temp;
}

$a = 1;
$b = 2;
do trySwap ($a, $b);
print $a;
print $b;
```
This Perl 1 program prints 12 which is different from what we expect.

### Call-by-reference in Perl 5
Call-by-reference is the default Perl 5 mechanism, just like the handbook states.
```
sub swap {
    $temp = $_[0];
    $_[0] = $_[1];
    $_[1] = $temp;
}

$a = 1;
$b = 2;
swap ($a, $b);
print $a;
print $b;
```
It prints 21 just like expected.

### Something similar in Perl 1
Here's a traditional although sophisticated Hello World program.
```
#!./perl

sub printSwapped {
    $temp = $_[0];
    $_[0] = $_[1];
    $_[1] = $temp;
    print $_[0];
    print $_[1];
}

$a = "world\n";
$b = "hello, ";
do printSwapped( $a, $b );
```
Its output is "hello, world" printed in a single line.
